glean-core/android/src/main/java/mozilla/telemetry/glean/utils/DateUtils.kt

```diff
@@ -0,0 +1,102 @@
+    TimeUnit.Minute to "yyyy-MM-dd'T'HH:mmZ",
+    TimeUnit.Hour to "yyyy-MM-dd'T'HHZ",
+    TimeUnit.Day to "yyyy-MM-ddZ"
+)
+
+@Suppress("TopLevelPropertyNaming")
+internal val DATE_FORMAT_PATTERN_VALUES = DATE_FORMAT_PATTERNS.values.toSet()
+
+/**
+ * Generate an ISO8601 compliant time string for the given time.
+ *
+ * @param date the [Date] object to convert to string
+ * @param truncateTo The TimeUnit to truncate the value to
+ * @return a string containing the date, time and timezone offset
+ */
+internal fun getISOTimeString(
+    date: Date = Date(),
+    truncateTo: TimeUnit = TimeUnit.Minute
+): String {
+    val cal = Calendar.getInstance()
+    cal.setTime(date)
+    return getISOTimeString(cal, truncateTo)
+}
+
+/**
+ * Generate an ISO8601 compliant time string for the given time.
+ *
+ * @param calendar the [Calendar] object to convert to string
+ * @param truncateTo The TimeUnit to truncate the value to
+ * @return a string containing the date, time and timezone offset
+ */
+internal fun getISOTimeString(
+    calendar: Calendar,
+    truncateTo: TimeUnit = TimeUnit.Minute
+): String {
+    val dateFormat = SimpleDateFormat(DATE_FORMAT_PATTERNS[truncateTo], Locale.US)
+    dateFormat.setTimeZone(calendar.getTimeZone())
+    val timeString = StringBuilder(dateFormat.format(calendar.getTime()))
+
+    // Due to limitations of SDK version 21, there isn't a way to properly output the time
+    // offset with a ':' character:
+    // 2018-12-19T12:36:00-0600    -- This is what we get
+    // 2018-12-19T12:36:00-06:00   -- This is what GCP will expect
+    //
+    // In order to satisfy time offset requirements of GCP, we manually insert the ":"
+    timeString.insert(timeString.length - 2, ":")
+
+    return timeString.toString()
+}
+
+/**
+ * Parses the subset of ISO8601 datetime strings generated by [getISOTimeString].
+ *
+ * Always returns the result in the device's current timezone offset, regardless of the
+ * timezone offset specified in the string.
+ *
+ * @param date a [String] representing an ISO date string generated by [getISOTimeString]
+ * @return a [Date] object representation of the provided string
+ */
+@Suppress("MagicNumber")
+internal fun parseISOTimeString(date: String): Date? {
+    // Due to limitations of SDK version 21, there isn't a way to properly parse the time
+    // offset with a ':' character:
+    // 2018-12-19T12:36:00-06:00  -- This is what we store
+    // 2018-12-19T12:36:00-0600   -- This is what SimpleDateFormat will expect
+
+    val correctedDate = if (date.get(date.length - 3) == ':') {
+        date.substring(0, date.length - 3) + date.substring(date.length - 2)
+    } else {
+        date
+    }
+
+    for (format in DATE_FORMAT_PATTERN_VALUES) {
+        val dateFormat = SimpleDateFormat(format, Locale.US)
+        try {
+            return dateFormat.parse(correctedDate)
+        } catch (e: java.text.ParseException) {
+            continue
+        }
+    }
+
+    return null
+}
```

Comment:
This feels like a lot of fiddly code to go outside of the Rust core.  On the other hand, it's fairly well tested in glean-ac, so it's fine to go in now.  Maybe file a bug?  It won't really matter until we have to duplicate this in Swift and C++ and ...

Reply:
You have a point. However, right now, this is only needed for the `parseISOTimeString` which is used by the `testGetValue(): Date`, which is platform/language specific, that's why I didn't port it over (I did port over `getISOTimeString` though). I thought about removing it from the file, but I wanted to keep changes to the bare minimum for now. Filed [bug 1556651](https://bugzilla.mozilla.org/show_bug.cgi?id=1556651) for tracking this.